questionpy/_util.py

```diff
@@ -1,11 +1,58 @@
+from collections.abc import Callable
 from types import UnionType
-from typing import TypeVar, get_args, get_type_hints
+from typing import Generic, TypeVar, cast, get_args, get_type_hints
 
-_T = TypeVar("_T", bound=type)
+_TypeT = TypeVar("_TypeT", bound=type)
+_T = TypeVar("_T")
+_UNSET = object()
 
 
-def get_mro_type_hint(klass: type, attr_name: str, bound: _T) -> _T:
-    # FIXME: This function is called too early, when the classes referenced in forward refs may not be defined yet.
+class _CachedClassProperty(Generic[_TypeT, _T]):
+    """See [cached_class_property][]."""
+
+    def __init__(self, getter: Callable[[_TypeT], _T]) -> None:
+        self._getter = getter
+        self._name: str | None = None
+
+    def __get__(self, instance: None, owner: _TypeT) -> _T:
+        if not self._name:
+            return self._getter(owner)
+
+        cached_value = owner.__dict__.get(self._name, _UNSET)
+        if cached_value is _UNSET:
+            value = self._getter(owner)
+            setattr(owner, self._name, value)
+            return value
+
+        return cached_value
```

Comment:
Vielleicht könnte man hier noch Kommentare schreiben, wie das genau funktioniert. Z.B., falls das so richtig ist:

```suggestion
    def __get__(self, instance: None, owner: _TypeT) -> _T:
        if not self._name:
            # The class was not instantiated as self.__set_name__ was not called yet.
            return self._getter(owner)

        cached_value = owner.__dict__.get(self._name, _UNSET)
        if cached_value is _UNSET:
            value = self._getter(owner)
            # Sets properties value and updates the read-only owner.__dict__.
            setattr(owner, self._name, value)
            return value

        return cached_value
```

Reply:
Guter Punkt. 3d8ecabec2592f19bc2779fad0c952f3092429ba